nifi-nar-bundles/nifi-framework-bundle/nifi-framework-external-resource-utils/src/main/java/org/apache/nifi/flow/resource/ExternalResourceListAsHtmlParser.java

```diff
@@ -0,0 +1,103 @@
+ * limitations under the License.
+ */
+package org.apache.nifi.flow.resource;
+
+import org.w3c.dom.Document;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.InputSource;
+import org.xml.sax.SAXException;
+
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.xpath.XPath;
+import javax.xml.xpath.XPathConstants;
+import javax.xml.xpath.XPathExpression;
+import javax.xml.xpath.XPathExpressionException;
+import javax.xml.xpath.XPathFactory;
+import java.io.IOException;
+import java.io.StringReader;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.Collection;
+
+public class ExternalResourceListAsHtmlParser {
+    private static final String DEFAULT_X_PATH_IDENTIFIER_FOR_FILE_LIST = "//tr[count(td)>2]";
+    private static final String DEFAULT_X_PATH_IDENTIFIER_FOR_LOCATION = "./td[1]/a/text()";
+    private static final String DEFAULT_X_PATH_IDENTIFIER_FOR_LAST_MODIFIED = "./td[2]/text()";
+    private static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("yyy-MM-dd HH:mm");
+
+    private final XPathExpression xPathIdentifierForFileList;
+    private final XPathExpression xPathIdentifierForLocation;
+    private final XPathExpression xPathIdentifierForLastModified;
+
+    public ExternalResourceListAsHtmlParser() throws XPathExpressionException {
+        final XPath xPath = XPathFactory.newInstance().newXPath();
+        xPathIdentifierForFileList = xPath.compile(DEFAULT_X_PATH_IDENTIFIER_FOR_FILE_LIST);
+        xPathIdentifierForLocation = xPath.compile(DEFAULT_X_PATH_IDENTIFIER_FOR_LOCATION);
+        xPathIdentifierForLastModified = xPath.compile(DEFAULT_X_PATH_IDENTIFIER_FOR_LAST_MODIFIED);
+    }
+
+    public Collection<ExternalResourceDescriptor> parseResponse(final String response, final String url) throws ParserConfigurationException, XPathExpressionException, IOException, SAXException {
+        final Collection<ExternalResourceDescriptor> result = new ArrayList<>();
+        final NodeList nodeList = gatherResourceNodeList(response);
+        for (int i = 0; i < nodeList.getLength(); i++) {
+            final Node node = nodeList.item(i);
+            if (isFile(node)) {
+                final ExternalResourceDescriptor descriptor = createDescriptor(node, url, true);
+                result.add(descriptor);
+            } else {
+                final ExternalResourceDescriptor descriptor = createDescriptor(node, url, false);
+                result.add(descriptor);
+            }
+        }
+        return result;
+    }
+
+    private NodeList gatherResourceNodeList(final String response) throws ParserConfigurationException, XPathExpressionException, IOException, SAXException {
+        final Document doc = DocumentBuilderFactory.newInstance().newDocumentBuilder().parse(new InputSource(new StringReader(response)));
```

Comment:
The default configuration of the `DocumentBuilderFactory` is vulnerable to XML External Entity processing. All XML handling should leverage components from `nifi-xml-processing`.